src/btree/bt_vrfy.c

```diff
@@ -582,6 +583,115 @@ __verify_addr_ts(WT_SESSION_IMPL *session, WT_REF *ref, WT_CELL_UNPACK *unpack,
+        WT_ERR_NOTFOUND_OK(__wt_config_getones(session, meta_value, "id", &id));
+        if (ret == WT_NOTFOUND) {
+            continue;
+        }
+        if (errno != ERANGE && btree_id == __wt_strtouq(id.str, NULL, 10)) {
+            meta_cursor->get_key(meta_cursor, uri);
+            goto done;
+        }
+    }
+
+done:
+err:
+    WT_TRET(__wt_metadata_cursor_release(session, &meta_cursor));
+    return (ret);
+}
+
+/*
+ * __wt_verify_history_store_tree --
+ *     Verify the history store tree, by iterating and checking all the keys in the history and that
+ *     it exists in both the meta file and the respective data store table. Skips duplicate keys by
+ *     storing the previous key.
+ */
+int
+__wt_verify_history_store_tree(WT_SESSION_IMPL *session)
+{
+    WT_CURSOR *cursor, *data_cursor;
+    WT_DECL_RET;
+    WT_ITEM hs_key;
+    WT_ITEM prev_hs_key;
+    WT_TIME_PAIR hs_start, hs_stop;
+    uint32_t prev_btree_id;
+    uint32_t session_flags, btree_id;
+    int cmp;
+    const char *uri;
+    bool first;
+
+    session_flags = 0;
+    /* Dodging uninitialized build error. Don't trust this initial value. */
+    prev_btree_id = 0;
+    uri = NULL;
+    first = true;
+
+    WT_ERR(__wt_hs_cursor(session, &session_flags));
+    cursor = session->hs_cursor;
+
+    while ((ret = cursor->next(cursor)) == 0) {
+        WT_ERR(cursor->get_key(cursor, &btree_id, &hs_key, &hs_start.timestamp, &hs_start.txnid,
+          &hs_stop.timestamp, &hs_stop.txnid));
+
+        /*
+         *  Keep track of the previous comparison. The history store is stored in order, so we can
+         *  avoid redundant comparisons. The first flag is used to avoid comparing against
+         *  uninitialized variables.
+         */
+        if (!first && prev_btree_id == btree_id) {
+            WT_ERR(__wt_compare(session, NULL, &hs_key, &prev_hs_key, &cmp));
+            if (cmp == 0) {
+                continue;
+            }
+        }
+
+        first = false;
+        WT_ERR(__wt_buf_initsize(session, &prev_hs_key, hs_key.size));
+        WT_ERR(__wt_buf_set(session, &prev_hs_key, hs_key.data, hs_key.size));
+        prev_btree_id = btree_id;
+
+        /*
+         * Find the URI from the metadata and validate the btree ID. Using this URI, verify the
+         * history store key with the data store.
+         */
+        WT_ERR(__verify_btree_id_with_meta(session, btree_id, &uri));
+        WT_ERR(__wt_open_cursor(session, uri, NULL, NULL, &data_cursor));
+
+        data_cursor->set_key(data_cursor, &hs_key);
+        WT_ERR_NOTFOUND_OK(data_cursor->search(data_cursor));
+        if (ret == WT_NOTFOUND) {
+            WT_ERR_MSG(session, WT_ERROR, "key exists in history store but not in data store");
+        }
+        WT_TRET(data_cursor->close(data_cursor));
+    }
+err:
+    WT_TRET(__wt_hs_cursor_close(session, session_flags));
+
+    if (ret != WT_NOTFOUND) {
+        return (ret);
+    }
+
+    return (0);
```

Comment:
braces aren't needed.
More importantly, I see what you're doing, but this means that `WT_NOTFOUND` from any called function can end up returning 0, that is, if `__wt_hs_cursor` returns `WT_NOTFOUND`, this function returns 0 and that's not OK. I'd suggest moving the `WT_NOTFOUND is OK` check to a specific point in the code, something like this:
```
        }
        WT_ERR(data_cursor->close(data_cursor)); 
    }
    WT_ERR_NOTFOUND_OK(ret);

err:
    WT_TRET(__wt_hs_cursor_close(session, session_flags));
    return (ret);
```